README.md
# another_dashed_container

A Dashed Container plugin for Flutter, easy to implement dashed line for your widgets

## Demo
![Demo: ](screenshot.png)

## Usage
To use plugin, just import package `import 'package:another_dashed_container/another_dashed_container.dart';`

## Example
You can check example directory to know how to use it like the demo image.

```dart
    import 'package:flutter/material.dart';
    import 'package:another_dashed_container/another_dashed_container.dart';

    void main() => runApp(MyApp());

    class MyApp extends StatelessWidget {
      @override
      Widget build(BuildContext context) {
        return MaterialApp(
          title: 'Flutter Demo',
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: MyHomePage(title: 'Flutter Demo Home Page'),
        );
      }
    }

    class MyHomePage extends StatefulWidget {
      MyHomePage({Key key, this.title}) : super(key: key);

      final String title;

      @override
      _MyHomePageState createState() => _MyHomePageState();
    }

    class _MyHomePageState extends State<MyHomePage> {
      @override
      Widget build(BuildContext context) {
        return Scaffold(
          appBar: AppBar(title: Text(widget.title)),
          body: SingleChildScrollView(
            child: Center(
              child: Column(
                mainAxisSize: MainAxisSize.max,
                crossAxisAlignment: CrossAxisAlignment.center,
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Text('Line container - blankLength = 0'),
                  _buildSpace(),
                  DashedContainer(
                    child: Container(
                      height: 200.0,
                      width: 200.0,
                      decoration: BoxDecoration(color: Colors.blue, borderRadius: BorderRadius.circular(10.0)),
                    ),
                    dashColor: Colors.black,
                    borderRadius: 10.0,
                    dashedLength: 30.0,
                    blankLength: 0.0,
                    strokeWidth: 6.0,
                  ),
                  _buildSpace(),
                  Text('Dashed container - rectangle'),
                  _buildSpace(),
                  DashedContainer(
                    child: Container(
                      height: 100.0,
                      width: 200.0,
                      decoration: BoxDecoration(color: Colors.red, borderRadius: BorderRadius.circular(20.0)),
                    ),
                    dashColor: Colors.black,
                    borderRadius: 20.0,
                    dashedLength: 30.0,
                    blankLength: 6.0,
                    strokeWidth: 6.0,
                  ),
                  _buildSpace(),
                  Text('Dashed container - circle'),
                  _buildSpace(),
                  DashedContainer(
                    child: Container(
                      height: 200.0,
                      width: 200.0,
                      decoration: BoxDecoration(color: Colors.orange, shape: BoxShape.circle),
                    ),
                    dashColor: Colors.black,
                    boxShape: BoxShape.circle,
                    dashedLength: 30.0,
                    blankLength: 6.0,
                    strokeWidth: 6.0,
                  ),
                  _buildSpace(),
                  Text('Dashed container - custom'),
                  _buildSpace(),
                  DashedContainer(
                    child: Padding(
                      padding: const EdgeInsets.all(20.0),
                      child: FlutterLogo(
                        size: 80,
                      ),
                    ),
                    dashColor: Colors.black,
                    boxShape: BoxShape.circle,
                    dashedLength: 30.0,
                    blankLength: 6.0,
                    strokeWidth: 6.0,
                  ),
                  _buildSpace(),
                ],
              ),
            ),
          ),
        );
      }

      Widget _buildSpace() {
        return SizedBox(height: 10.0);
      }
    }

```


CHANGELOG.md
## [1.0.0]
* initial release.


lib/dash_offset.dart
part of another_dashed_container;

enum _DashOffsetType { Absolute, Percentage }

/// Specifies the starting position of a dash array on a path, either as a
/// percentage or absolute value.
///
/// The internal value will be guaranteed to not be null.
class DashOffset {
  /// Create a DashOffset that will be measured as a percentage of the length
  /// of the segment being dashed.
  ///
  /// `percentage` will be clamped between 0.0 and 1.0; null will be converted
  /// to 0.0.
  DashOffset.percentage(double percentage)
      : _rawVal = percentage.clamp(0.0, 1.0),
        _dashOffsetType = _DashOffsetType.Percentage;

  /// Create a DashOffset that will be measured in terms of absolute pixels
  /// along the length of a [Path] segment.
  ///
  /// `start` will be coerced to 0.0 if null.
  const DashOffset.absolute(double start)
      : _rawVal = start,
        _dashOffsetType = _DashOffsetType.Absolute;

  final double _rawVal;
  final _DashOffsetType _dashOffsetType;

  double _calculate(double length) {
    return _dashOffsetType == _DashOffsetType.Absolute ? _rawVal : length * _rawVal;
  }
}

/// A circular array of dash offsets and lengths.
///
/// For example, the array `[5, 10]` would result in dashes 5 pixels long
/// followed by blank spaces 10 pixels long.  The array `[5, 10, 5]` would
/// result in a 5 pixel dash, a 10 pixel gap, a 5 pixel dash, a 5 pixel gap,
/// a 10 pixel dash, etc.
///
/// Note that this does not quite conform to an [Iterable<T>], because it does
/// not have a moveNext.
class CircularIntervalList<T> {
  CircularIntervalList(this._vals);

  final List<T> _vals;
  int _idx = 0;

  T get next {
    if (_idx >= _vals.length) {
      _idx = 0;
    }
    return _vals[_idx++];
  }
}


lib/dash_path_painter.dart
part of another_dashed_container;

class DashPathPainter extends CustomPainter {
  final Paint mPaint;
  final double? blankWidth;
  final double? dashedWidth;
  final double? borderRadius;
  final BoxShape? boxShape;

  DashPathPainter({
    required this.mPaint,
    this.blankWidth,
    this.dashedWidth,
    this.borderRadius,
    this.boxShape,
  });

  @override
  bool shouldRepaint(DashPathPainter oldDelegate) => true;

  @override
  void paint(Canvas canvas, Size size) {
    Path p;
    double? bRadius = borderRadius;

    if (boxShape == BoxShape.rectangle) {
      if (bRadius == 0)
        p = Path()..addRect(Rect.fromLTWH(0.0, 0.0, size.width, size.height));
      else {
        Radius radius = Radius.circular(bRadius!);
        p = Path()
          ..addRRect(
            RRect.fromRectAndCorners(
              Rect.fromLTWH(0.0, 0.0, size.width, size.height),
              bottomLeft: radius,
              bottomRight: radius,
              topLeft: radius,
              topRight: radius,
            ),
          );
      }
    } else {
      bRadius = size.width / 2;
      Radius radius = Radius.circular(bRadius);
      p = Path()
        ..addRRect(
          RRect.fromRectAndCorners(
            Rect.fromLTWH(0.0, 0.0, size.width, size.height),
            bottomLeft: radius,
            bottomRight: radius,
            topLeft: radius,
            topRight: radius,
          ),
        );
    }
    canvas.drawPath(
      dashPath(
        p,
        dashArray: CircularIntervalList<double?>(
          <double?>[
            dashedWidth,
            blankWidth,
          ],
        ),
      ),
      mPaint,
    );
  }
}

/// Creates a new path that is drawn from the segments of `source`.
///
/// Dash intervals are controled by the `dashArray` - see [CircularIntervalList]
/// for examples.
///
/// `dashOffset` specifies an initial starting point for the dashing.
///
/// Passing in a null `source` will result in a null result.  Passing a `source`
/// that is an empty path will return an empty path.
Path dashPath(
  Path source, {
  required CircularIntervalList<double?> dashArray,
  DashOffset? dashOffset,
}) {
  assert(dashArray != null);

  dashOffset = dashOffset ?? const DashOffset.absolute(0.0);
  // TODO: Is there some way to determine how much of a path would be visible today?

  final Path dest = Path();
  for (final PathMetric metric in source.computeMetrics()) {
    double distance = dashOffset._calculate(metric.length);
    bool draw = true;
    while (distance < metric.length) {
      final double len = dashArray.next!;
      if (draw) {
        dest.addPath(metric.extractPath(distance, distance + len), Offset.zero);
      }
      distance += len;
      draw = !draw;
    }
  }

  return dest;
}


example/lib/main.dart
import 'package:another_dashed_container/another_dashed_container.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.title)),
      body: SingleChildScrollView(
        child: Center(
          child: Column(
            mainAxisSize: MainAxisSize.max,
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text('Line container - blankLength = 0'),
              _buildSpace(),
              DashedContainer(
                child: Container(
                  height: 200.0,
                  width: 200.0,
                  decoration: BoxDecoration(color: Colors.blue, borderRadius: BorderRadius.circular(10.0)),
                ),
                dashColor: Colors.black,
                borderRadius: 10.0,
                dashedLength: 30.0,
                blankLength: 0.0,
                strokeWidth: 6.0,
              ),
              _buildSpace(),
              Text('Dashed container - rectangle'),
              _buildSpace(),
              DashedContainer(
                child: Container(
                  height: 100.0,
                  width: 200.0,
                  decoration: BoxDecoration(color: Colors.red, borderRadius: BorderRadius.circular(20.0)),
                ),
                dashColor: Colors.black,
                borderRadius: 20.0,
                dashedLength: 30.0,
                blankLength: 6.0,
                strokeWidth: 6.0,
              ),
              _buildSpace(),
              Text('Dashed container - circle'),
              _buildSpace(),
              DashedContainer(
                child: Container(
                  height: 200.0,
                  width: 200.0,
                  decoration: BoxDecoration(color: Colors.orange, shape: BoxShape.circle),
                ),
                dashColor: Colors.black,
                boxShape: BoxShape.circle,
                dashedLength: 30.0,
                blankLength: 6.0,
                strokeWidth: 6.0,
              ),
              _buildSpace(),
              Text('Dashed container - custom'),
              _buildSpace(),
              DashedContainer(
                child: Padding(
                  padding: const EdgeInsets.all(20.0),
                  child: FlutterLogo(
                    size: 80,
                  ),
                ),
                dashColor: Colors.black,
                boxShape: BoxShape.circle,
                dashedLength: 30.0,
                blankLength: 6.0,
                strokeWidth: 6.0,
              ),
              _buildSpace(),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildSpace() {
    return SizedBox(height: 10.0);
  }
}
